File: src/Tests/Behavioral/ReverseSortNaNOrder/ReverseSortNaNOrder.go
package main

import (
	"fmt"
	"math"
)

// Interface mirrors sort.Interface — the dispatch under test flows through it.
type Interface interface {
	Len() int
	Less(i, j int) bool
	Swap(i, j int)
}

// Float64Slice mirrors sort.Float64Slice: a NaN-aware Less over concrete float64.
type Float64Slice []float64

func (x Float64Slice) Len() int { return len(x) }
func (x Float64Slice) Less(i, j int) bool {
	return x[i] < x[j] || (fisNaN(x[i]) && !fisNaN(x[j]))
}
func (x Float64Slice) Swap(i, j int) { x[i], x[j] = x[j], x[i] }

// fisNaN is sort's local isNaN: concrete float64 self-inequality.
func fisNaN(f float64) bool { return f != f }

// reverse mirrors sort's reverse: an embedded Interface with a Less OVERRIDE.
// The override must win over the promoted embedded method through interface dispatch.
type reverse struct {
	Interface
}

func (r reverse) Less(i, j int) bool {
	return r.Interface.Less(j, i)
}

func Reverse(data Interface) Interface {
	return &reverse{data}
}

// insertionSort dispatches Len/Less/Swap through the interface value.
func insertionSort(data Interface) {
	n := data.Len()
	for i := 1; i < n; i++ {
		for j := i; j > 0 && data.Less(j, j-1); j-- {
			data.Swap(j, j-1)
		}
	}
}

// ordered mirrors cmp.Ordered's operator surface for the generic legs.
type ordered interface {
	~int | ~float32 | ~float64 | ~string
}

// isNaN mirrors cmp.isNaN: GENERIC self-inequality — the emission routes through
// golib equality, which must follow Go's IEEE `==` (NaN unequal to itself).
func isNaN[T ordered](x T) bool { return x != x }

// less mirrors cmp.Less: NaN orders before any non-NaN.
func less[T ordered](x, y T) bool { return (isNaN(x) && !isNaN(y)) || x < y }

// eq exercises the plain `comparable` generic equality path.
func eq[T comparable](a, b T) bool { return a == b }

func main() {
	nan := math.NaN()

	// Reverse (embed-override) dispatch: descending sort through Reverse.
	ints := Float64Slice{74, 59, 238, -784, 9845, 959, 905, 0, 42}
	insertionSort(Reverse(ints))
	fmt.Println(ints)

	// NaN-aware ascending sort through the concrete Less.
	floats := Float64Slice{74.3, 59.0, math.Inf(1), 238.2, -784.0, 2.3, nan, nan, math.Inf(-1), 905}
	insertionSort(floats)
	fmt.Println(floats)

	// Generic NaN probes (the cmp.isNaN / cmp.Less shapes).
	fmt.Println(isNaN(nan), isNaN(1.5), isNaN(float32(math.NaN())), isNaN(7), isNaN("x"))
	fmt.Println(less(nan, -784.0), less(-784.0, nan), less(nan, nan), less(1.0, 2.0))

	// Generic `comparable` equality: IEEE semantics for every float-bearing kind.
	fmt.Println(eq(nan, nan), eq(1.5, 1.5))
	fmt.Println(eq(float32(math.NaN()), float32(math.NaN())), eq(float32(1.5), float32(1.5)))
	fmt.Println(eq(complex(nan, 0), complex(nan, 0)), eq(complex(1, 2), complex(1, 2)))
	fmt.Println(eq(complex64(complex(nan, 0)), complex64(complex(nan, 0))), eq(complex64(complex(1, 2)), complex64(complex(1, 2))))

	// Boxed (interface) equality: Go compares dynamic values with the type's `==`.
	var bx any = nan
	var by any = nan
	fmt.Println(bx == by, any(1.5) == any(1.5))
}

File: src/Tests/Behavioral/ReverseSortNaNOrder/ReverseSortNaNOrder.cs
namespace go;

using fmt = fmt_package;
using Δmath = math_package;

partial class main_package {

[GoType] partial interface Interface {
    nint Len();
    bool Less(nint i, nint j);
    void Swap(nint i, nint j);
}

[GoType("[]float64")] partial struct Float64Slice;

public static nint Len(this Float64Slice x) {
    return len(x);
}

public static bool Less(this Float64Slice x, nint i, nint j) {
    return x[i] < x[j] || (fisNaN(x[i]) && !fisNaN(x[j]));
}

public static void Swap(this Float64Slice x, nint i, nint j) {
    (x[i], x[j]) = (x[j], x[i]);
}

internal static bool fisNaN(float64 f) {
    return f != f;
}

[GoType] partial struct reverse {
    public Interface Interface;
}

internal static bool Less(this reverse r, nint i, nint j) {
    return r.Interface.Less(j, i);
}

public static Interface Reverse(Interface data) {
    return new reverseжInterface(Ꮡ(new reverse(data)));
}

internal static void insertionSort(Interface data) {
    nint n = data.Len();
    for (nint i = 1; i < n; i++) {
        for (nint j = i; j > 0 && data.Less(j, j - 1); j--) {
            data.Swap(j, j - 1);
        }
    }
}

[GoType("operators = Sum, Comparable, Ordered")]
partial interface ordered<ΔT> {
    //  Type constraints: ~int | ~float32 | ~float64 | ~string
    // Derived operators: +, ==, !=, <, <=, >, >=
}

internal static bool isNaN<T>(T x)
    where T : /* ordered */ IAdditionOperators<T, T, T>, IEqualityOperators<T, T, bool>, IComparisonOperators<T, T, bool>, new()
{
    return !AreEqual(x, x);
}

internal static bool less<T>(T x, T y)
    where T : /* ordered */ IAdditionOperators<T, T, T>, IEqualityOperators<T, T, bool>, IComparisonOperators<T, T, bool>, new()
{
    return (isNaN(x) && !isNaN(y)) || x < y;
}

internal static bool eq<T>(T a, T b)
    where T : /* comparable */ new()
{
    return AreEqual(a, b);
}

internal static void Main() {
    var nan = Δmath.NaN();
    var ints = new Float64Slice(new float64[]{74D, 59D, 238D, -784D, 9845D, 959D, 905D, 0D, 42D}.slice());
    insertionSort(Reverse(ints));
    fmt.Println(ints);
    var floats = new Float64Slice(new float64[]{74.3D, 59.0D, Δmath.Inf(1), 238.2D, -784.0D, 2.3D, nan, nan, Δmath.Inf(-1), 905D}.slice());
    insertionSort(floats);
    fmt.Println(floats);
    fmt.Println(isNaN(nan), isNaN(1.5D), isNaN((float32)Δmath.NaN()), isNaN(7), isNaN((@string)"x"));
    fmt.Println(less(nan, -784.0D), less(-784.0D, nan), less(nan, nan), less(1.0D, 2.0D));
    fmt.Println(eq(nan, nan), eq(1.5D, 1.5D));
    fmt.Println(eq((float32)Δmath.NaN(), (float32)Δmath.NaN()), eq((float32)1.5F, (float32)1.5F));
    fmt.Println(eq(complex(nan, 0D), complex(nan, 0D)), eq(complex(1D, 2D), complex(1D, 2D)));
    fmt.Println(eq((complex64)complex(nan, 0D), (complex64)complex(nan, 0D)), eq((complex64)complex(1F, 2F), (complex64)complex(1F, 2F)));
    any bx = nan;
    any by = nan;
    fmt.Println(AreEqual(bx, by), AreEqual(((any)1.5D), ((any)1.5D)));
}

} // end main_package
